localstack/http/websocket.py

```diff
@@ -0,0 +1,326 @@
+
+    pass
+
+
+class WebSocketDisconnectedError(WebSocketError):
+    """Raised when the client has disconnected while the server is still trying to receive data."""
+
+    default_code = 1005
+    """https://asgi.readthedocs.io/en/latest/specs/www.html#disconnect-receive-event-ws"""
+
+    def __init__(self, code: int = None):
+        self.code = code if code is not None else self.default_code
+        super().__init__(f"Websocket disconnected code={self.code}")
+
+
+class WebSocketProtocolError(WebSocketError):
+    """Raised if there is a problem in the interaction between app and the websocket server."""
+
+    pass
+
+
+class WebSocket:
+    """
+    High-level interface to interact with a websocket after a handshake has been completed with
+    `WebsocketRequest.accept()`.
+    """
+
+    request: "WebSocketRequest"
+    socket: ASGIWebSocket
+    encoding: str = "utf-8"
+
+    def __init__(self, request: "WebSocketRequest", socket: ASGIWebSocket):
+        self.request = request
+        self.socket = socket
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def __iter__(self):
+        while True:
+            try:
+                yield self.receive()
+            except WebSocketDisconnectedError:
+                break
+
+    def send(self, text_or_bytes: str | bytes, timeout: float = None):
+        """
+        Send data to the websocket connection.
+
+        :param text_or_bytes: the data to send. Use strings for text-mode sockets (default).
+        :param timeout: the timeout in seconds to wait before raising a timeout error
+        """
+        if isinstance(text_or_bytes, str):
+            self.socket.send(
+                {
+                    "type": "websocket.send",
+                    "bytes": None,
+                    "text": text_or_bytes,
+                },
+                timeout=timeout,
+            )
+        else:
+            self.socket.send(
+                {
+                    "type": "websocket.send",
+                    "bytes": text_or_bytes,
+                    "text": None,
+                },
+                timeout=timeout,
+            )
+
+    def receive(self) -> str:
+        """
+        Receive text data from the websocket. This will automatically decode the data if the websocket is in
+        binary mode.
+
+        :raise WebSocketDisconnectedError: if the websocket was closed in the meantime
+        :raise WebSocketProtocolError: error in the interaction between the app and the webserver
+        :return: the next data package from the websocket
+        """
+        event = self.socket.receive()
+        if event["type"] == "websocket.receive":
+            text = event.get("text")
+            if text is not None:
+                return text
+
+            buf = event.get("bytes")
+            if buf is not None:
+                return buf.decode(self.encoding)
```

Comment:
nit: I think this would break if the sent bytes are not UTF-8 encoded / aren't encoded text (but this function clearly expects a string being sent, so maybe this is on purpose?).

Reply:
good point, i haven't really tested binary mode at all, and the API doesn't support it well as you pointed out. will test this explicitly and change the api to return either `str|bytes` depending on what mode the socket is in.